### ISSUE TITLE
[Refactor] Merge `LeagueToolkit.IO.Extensions` with the main project

### ISSUE THREAD

**Issue by Crauzer** [LoL-Fantome/LeagueToolkit#179]
I will be doing this change due to difficulty of maintaining/doing CI/CD for a .NET monorepo.

It's possible that the project might be split into several packages in the future, but currently it doesn't make much sense except for meta classes.

**Comment by Morilli** [comment]
The main rationale behind having the `IO.Extensions` part split was that other projects wouldn't need to pull in the reasonably large `Sixlabors.Imagesharp` dependency when they don't need it.

**Comment by Crauzer** [comment]
I will be refactoring the majority of this library and in the process deprecating that library.
For now I want to merge it into the main project so it can be easier to work on.
Besides that I also don't think the image library is that big, the assembly is 1.7 MB. I plan to use it much more in some new features that I want to implement.

**Comment by Rerago** [comment]
My goal back then (#134) was to keep the main project smaller.

This is exactly the "problem" i currently have. With my application ([LoLWideScreenFix](https://github.com/Rerago/LoLWideScreenFix)) I have a release output of 46mb. And these are to very large parts to `LeagueToolkit`, which would be so far ok if I would need it. But I made a branch to show the difference:

Full `LeagueToolkit` (without Extensions) => ~46mb
https://github.com/Rerago/LoLWideScreenFix/commit/37652b81a4d2447343546e0bb393a929d36063d5

Remove of `FlatSharp` (`IO.ReleaseManifestFile` is removed) => ~4,9mb
https://github.com/Rerago/LoLWideScreenFix/commit/b451921498186cfb55cc3ad132b96e77fdea3c2b

LeagueToolkitMini (the most necessary until my application was running) => 2,2mb
https://github.com/Rerago/LoLWideScreenFix/commit/e0068e92b702d3217802cbf437721af8acf1dba2

I think you can see quite well that the dependencies make up the bulk of the file size and I find that quite a pity. It is just a pity that one so e.g. `FlatSharp` drags along only for `IO.ReleaseManifestFile` (which would sound like a new nuget package to me, for example).

But i would be ok if `LeagueToolkit.IO.Extensions` would be integrated back into the main project for refactoring.

PS: I just wanted to share with the post that I would appreciate a split in a main project + many additional projects (with e.g. dependencies) very much (but I think that's what you intended with #181). Therefore, thank you for the great work on this project.

**Comment by Crauzer** [comment]
> My goal back then (#134) was to keep the main project smaller.
> 
> This is exactly the "problem" i currently have. With my application ([LoLWideScreenFix](https://github.com/Rerago/LoLWideScreenFix)) I have a release output of 46mb. And these are to very large parts to `LeagueToolkit`, which would be so far ok if I would need it. But I made a branch to show the difference:
> 
> Full `LeagueToolkit` (without Extensions) => ~46mb [Rerago/LoLWideScreenFix@37652b8](https://github.com/Rerago/LoLWideScreenFix/commit/37652b81a4d2447343546e0bb393a929d36063d5)
> 
> Remove of `FlatSharp` (`IO.ReleaseManifestFile` is removed) => ~4,9mb [Rerago/LoLWideScreenFix@b451921](https://github.com/Rerago/LoLWideScreenFix/commit/b451921498186cfb55cc3ad132b96e77fdea3c2b)
> 
> LeagueToolkitMini (the most necessary until my application was running) => 2,2mb [Rerago/LoLWideScreenFix@e0068e9](https://github.com/Rerago/LoLWideScreenFix/commit/e0068e92b702d3217802cbf437721af8acf1dba2)
> 
> I think you can see quite well that the dependencies make up the bulk of the file size and I find that quite a pity. It is just a pity that one so e.g. `FlatSharp` drags along only for `IO.ReleaseManifestFile` (which would sound like a new nuget package to me, for example).
> 
> But i would be ok if `LeagueToolkit.IO.Extensions` would be integrated back into the main project for refactoring.
> 
> PS: I just wanted to share with the post that I would appreciate a split in a main project + many additional projects (with e.g. dependencies) very much (but I think that's what you intended with #181). Therefore, thank you for the great work on this project.

Are the sizes you specified sizes of the compiled `LeagueToolkit.dll` ? If so, that'd mean something is very wrong :D
If it's the size of the compiled `.exe` then ~40mb is quite normal for .NET executables, at least that used to be the case with `.NET Core 3.1` and `.NET 5`. This seems like really strange behavior since the libraries you're removing are definitely not that big.

**Comment by Crauzer** [comment]
> PS: I just wanted to share with the post that I would appreciate a split in a main project + many additional projects (with e.g. dependencies) very much (but I think that's what you intended with #181). Therefore, thank you for the great work on this project.

The purpose of this PR and #181 is indeed to change the code organization of the whole project. This will take quite some time to do which is why I want to merge the library now. After the refactoring is done, I'd like to split the project into at least 2/3 different libraries (`LeagueToolkit.Core` - core functionality, `LeagueToolkit` - code for doing fancy stuff like conversion etc.. (includes `Core`), `LeagueToolkit.Meta` - BIN serialization)

**Comment by Rerago** [comment]
> Are the sizes you specified sizes of the compiled `LeagueToolkit.dll` ? If so, that'd mean something is very wrong :D If it's the size of the compiled `.exe` then ~40mb is quite normal for .NET executables, at least that used to be the case with `.NET Core 3.1` and `.NET 5`. This seems like really strange behavior since the libraries you're removing are definitely not that big. (Maybe it's the compiler/runtime optimizing away junk code ?)

See the following example project ([FlatSharpTest.zip](https://github.com/LoL-Fantome/LeagueToolkit/files/10299739/FlatSharpTest.zip))

- Simple console application that does nothing (except a `Console.WriteLine` to some `LeagueToolkit` class)
- Original `LeagueToolkit` (only slightly modified .csproj to build with and without FlatSharp)

If you run the `publish.bat` and look at the result it roughly mirrors the result with my application.

Relase = ~4,3mb
noFlatSharp = ~45,4mb

Whatever `FlatSharp` does on compile it takes MASSIVE space :/
(Without the `ReleaseManifestFile` being needed in any way)

What I don't have in the example project. If you build the application completely without `LeagueToolkit` it would be 150kb with the same publish command. So I wouldn't say that ~40mb is normal for a `no-self-contained` application.

**Comment by Crauzer** [comment]
> Whatever `FlatSharp` does on compile it takes MASSIVE space :/
> (Without the `ReleaseManifestFile` being needed in any way)

I think I found the culprit behind this issue, look at the libraries the project is using, they're **huge**. They're bundling the runtime, roslyn compiler toolkit and similar libraries for use in the `FlatSharp.Compiler` project which is used for compiling schemas.
https://github.com/jamescourtney/FlatSharp/blob/main/src/FlatSharp/FlatSharp.csproj#L30-L44

**Comment by Rerago** [comment]
So I just tried the new beta with #188. There was a partial success:

3.0.3 		= ~46.1mb
4.0.0-beta.2 	= ~36,7mb

I think only outsourcing would really help.

**Comment by Morilli** [comment]
I imagine the main "big" dependency is the `Microsoft.CodeAnalysis.CSharp` required by the MetaDump generation code.
You can check the specific dlls that are large in your output by not publishing as single file executable.

**Comment by Rerago** [comment]
Microsoft.CodeAnalysis.CSharp => 6mb
Microsoft.CodeAnalysis => 2,7mb

But the whole folder is only ~16mb. During the compile time, it looks like something is still being added.

**Comment by Morilli** [comment]
The increase from 16mb to 36 mb in your case is most likely caused by the `PublishReadyToRun` option.

**Comment by Rerago** [comment]
Yes well but from 16mb to 36mb that's 20mb. That's more than the DLL itself.

And completely without FlatSharp (IO.ReleaseManifestFile) the build is only 4.9mb.

**Comment by Morilli** [comment]
[The documentation of the ReadyToRun feature](https://learn.microsoft.com/en-us/dotnet/core/deploying/ready-to-run#impact-of-using-the-readytorun-feature) explicitly mentions that assembly size is massively increased:
`In general, the size of an assembly will grow to between two to three times larger.`

And yes, the `Microsoft.CodeAnalysis.CSharp` dependency is the biggest one currently, but on the current beta this is not longer required by `IO.ReleaseManifestFile`, but by `MetaDump`.

**Comment by Rerago** [comment]
You seem to be right. I have done the following:

````xml
<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.4.0" Condition="'$(Configuration)'=='DEBUG'" />
````

And the whole MetaDump.cs into a big `#IF DEBUG`.

The result looks much better =>5.9MB.

**Comment by Crauzer** [comment]
I will not be moving the Meta generation code into a `debug` conditional, I will however split it from the main project once I'm done with refactoring and I figure out how to do the devops for the project.

**Comment by Crauzer** [comment]
#234
